### PR TITLE
gh-101578: Amend exception docs

### PR DESCRIPTION
**Pull request by erlend-aasland**
Follow-up to gh-101962

* Issue: gh-101578

### CODE REVIEW

File: Doc/c-api/exceptions.rst [hunk 1]
```diff
@@ -438,17 +438,18 @@ Querying the error indicator
...
    set, it will be cleared and you own a reference to each object retrieved.  The
    value and traceback object may be ``NULL`` even when the type object is not.
 
    .. note::
 
-      This function is normally only used by code that needs to catch exceptions or
-      by code that needs to save and restore the error indicator temporarily, e.g.::
+      This function is normally only used by legacy code that needs to catch
+      exceptions or by code that needs to save and restore the error indicator
+      temporarily.
```
Comment (CAM-Gerlach) [inline Doc/c-api/exceptions.rst]:
```suggestion
      exceptions or save and restore the error indicator temporarily.
```

More concise, avoid repeating "code that needs to" here and ensure "legacy" applies to it too.

File: Doc/c-api/exceptions.rst [hunk 2]
```diff
@@ -461,10 +462,10 @@ Querying the error indicator
 
    .. deprecated:: 3.12
 
+      Use :c:func:`PyErr_GetRaisedException` instead.
```
Comment (CAM-Gerlach) [inline Doc/c-api/exceptions.rst]: I suggest moving this directive to the top of the `.. c:function` content block, like the original message was, so users have a heads-up right away that the function is deprecated and don't have to through the whole description only to find a notice that it is deprecated and they should use something else buried at the bottom.

**Comment by CAM-Gerlach** [comment]
(And in turn, followup to PR #101607)

File: Doc/c-api/exceptions.rst [hunk 3]
```diff
@@ -461,10 +462,10 @@ Querying the error indicator
...
 
-.. c:function:: void PyErr_Restore(PyObject *type, PyObject *value, PyObject *traceback)
 
-    As of 3.12, this function is deprecated. Use :c:func:`PyErr_SetRaisedException` instead.
+.. c:function:: void PyErr_Restore(PyObject *type, PyObject *value, PyObject *traceback)
 
    Set the error indicator from the three objects.  If the error indicator is
    already set, it is cleared first.  If the objects are ``NULL``, the error
```
Comment (CAM-Gerlach) [inline Doc/c-api/exceptions.rst]: Maybe worth porting the heavily revised/clarified text you implemented for `PyErr_SetRaisedException` in PR #101962 to here as well, given this text is very similar to nearly identical (being the source of that one's)?

File: Doc/c-api/exceptions.rst [hunk 4]
```diff
@@ -478,18 +479,16 @@ Querying the error indicator
...
-      This function is normally only used by code that needs to save and restore the
-      error indicator temporarily.  Use :c:func:`PyErr_Fetch` to save the current
-      error indicator.
+      This function is normally only used by legacy code that needs to
+      save and restore the error indicator temporarily.
+      Use :c:func:`PyErr_Fetch` to save the current error indicator.
 
    .. deprecated:: 3.12
 
+      Use :c:func:`PyErr_SetRaisedException` instead.
```
Comment (CAM-Gerlach) [inline Doc/c-api/exceptions.rst]: Same here—I suggest moving this to the top where the original message was.

File: Doc/c-api/exceptions.rst [hunk 5]
```diff
@@ -509,6 +508,9 @@ Querying the error indicator
 
    .. deprecated:: 3.12
 
+      Use :c:func:`PyErr_GetRaisedException` instead,
+      to avoid any possible de-normalization.
```
Comment (CAM-Gerlach) [inline Doc/c-api/exceptions.rst]: Likewise

**Review by CAM-Gerlach — COMMENTED** [review]
Thanks @erlend-aasland ! Just some relatively modest suggestions and comments.

File: Doc/c-api/exceptions.rst (hunk 3 above)
Comment (erlend-aasland) [reply Doc/c-api/exceptions.rst]: PTAL!

**Review by iritkatriel — APPROVED** [review]

File: Doc/c-api/exceptions.rst [hunk 6]
```diff
@@ -459,15 +463,17 @@ Querying the error indicator
...
 
-   Set the error indicator from the three objects.  If the error indicator is
-   already set, it is cleared first.  If the objects are ``NULL``, the error
+   Set the error indicator from the three objects,
+   *type*, *value*, and *traceback*,
+   clearing the existing exception if one is set.
+   If the objects are ``NULL``, the error
    indicator is cleared.  Do not pass a ``NULL`` type and non-``NULL`` value or
    traceback.  The exception type should be a class.  Do not pass an invalid
    exception type or value. (Violating these rules will cause subtle problems
```
Comment (CAM-Gerlach) [inline Doc/c-api/exceptions.rst:479]: I guess I was thinking of also revising this part along the lines of the previous, and programmatically marking the function as stealing a reference (if it isn't already), and perhaps with the warning message as in the previous. But up to you if you think it's worth it.

**Review by CAM-Gerlach — APPROVED** [review]
LGTM, other than if you want to further revise the `PyErr_SetRaisedException` section along your previous lines (per comment).

File: Doc/c-api/exceptions.rst (hunk 6 above)
Comment (erlend-aasland) [reply Doc/c-api/exceptions.rst:479]: Sure, sounds good to me.

Comment (erlend-aasland) [reply Doc/c-api/exceptions.rst:479]: Hm, I'm afraid we'll have to reword most of `PyErr_Fetch` and `PyErr_Restore` in this case. Let's land this so we can close the linked issue.

Further improvements and amendments can be made in small increments, using the _skip issue_ label.

**Pull request merged**